### PR TITLE
Add parallel scan 

### PR DESCRIPTION
**Pull request by Sudha247**

### CODE REVIEW

File: lib/task.ml [hunk 1]
```diff
@@ -81,3 +81,51 @@ let parallel_for_reduce pool reduce_fun init ~chunk_size ~start ~finish ~body =
...
+  let n = Array.length elements in
+  let p = (Array.length pool.domains) + 1 in (*total processing elements*)
+  let prefix_s = Array.copy elements in(*for having the same type as elements*)
+
+  (*sweep 1*)
+  parallel_for pool ~chunk_size:1 ~start:0 ~finish:(p - 1)
+  ~body:(fun i ->
+    let s = (i * n) / (p ) in
+    let e = (i + 1) * n / (p ) - 1 in
+    (* Printf.printf "s = %d e = %d\n" s e; *)
```
Comment (kayceesrk) [inline lib/task.ml:108]: Remove comments.

File: lib/task.mli [hunk 2]
```diff
@@ -39,3 +39,9 @@ val parallel_for_reduce : pool -> ('a -> 'a -> 'a) -> 'a -> chunk_size:int ->
 (** [parallel_for_reduce p r i c s f b] is similar to [parallel_for] except
   * that the result returned by each iteration is reduced with [r] with initial
   * value [i]. *)
+
+val parallel_scan : pool -> ('a -> 'a -> 'a) -> 'a array -> 'a array
+(** [parallel_scan p op a] computes the scan of the array a
+  * in parallel with binary operator op and returns the result array.
```
Comment (kayceesrk) [inline lib/task.mli:45]:
```suggestion
  * in parallel with binary operator [op] and returns the result array.
```

File: lib/task.mli [hunk 3]
```diff
@@ -39,3 +39,9 @@ val parallel_for_reduce : pool -> ('a -> 'a -> 'a) -> 'a -> chunk_size:int ->
 (** [parallel_for_reduce p r i c s f b] is similar to [parallel_for] except
   * that the result returned by each iteration is reduced with [r] with initial
   * value [i]. *)
+
+val parallel_scan : pool -> ('a -> 'a -> 'a) -> 'a array -> 'a array
+(** [parallel_scan p op a] computes the scan of the array a
```
Comment (kayceesrk) [inline lib/task.mli:44]:
```suggestion
(** [parallel_scan p op a] computes the scan of the array [a]
```

File: lib/task.mli [hunk 4]
```diff
@@ -39,3 +39,9 @@ val parallel_for_reduce : pool -> ('a -> 'a -> 'a) -> 'a -> chunk_size:int ->
...
+  * in parallel with binary operator op and returns the result array.
+  * Scan is similar to Array.fold_left but returns an array of reduced
```
Comment (kayceesrk) [inline lib/task.mli:46]:
```suggestion
  * Scan is similar to [Array.fold_left] but returns an array of reduced
```

**Comment by kayceesrk** [comment]
Can you comment on how this scales @Sudha247?

**Comment by Sudha247** [comment]
These are some numbers for the `prefix_sum` test:

| #Cores | time  | time (prefix_sum function) | Speedup  |
|--------|------|------------------|----------|
| 1      | 3.07 | 2.15             | 1.00     |
| 2      | 2.98 | 2.05             | 1.04     |
| 4      | 2.42 | 1.5              | 1.43     |
| 8      | 2.24 | 1.22             | 1.76     |
| 12     | 2.05 | 1.13             | 1.90     |
| 16     | 2.02 | 1.09             | 1.97     |
| 20     | 2.05 | 1.12             | 1.91     |
| 24     | 2.06 | 1.13             | 1.90     |

**Comment by kayceesrk** [comment]
Thanks. I'll merge this one. We should revisit the performance. I would be curious to know where the overheads lie.

**Pull request merged**